envoy/upstream/load_balancer.h

```diff
@@ -165,6 +165,8 @@ struct LoadBalancerParams {
   const PrioritySet& priority_set;
   // The worker local priority set of the local cluster.
   const PrioritySet* local_priority_set{};
+  // The main thread dispatcher, or nullptr if not specified.
+  Event::Dispatcher* main_thread_dispatcher = nullptr;
```

Comment:
why doesn't use reference here?

/wait-any

Reply:
Great question! This is unfortunate work around 2 issues:

1. I didn't find how to plumb the main_thread_dispatcher to `SubsetLoadBalancer::PrioritySubsetImpl::PrioritySubsetImpl()`

2. Many tests locally define`Upstream::LoadBalancerParams lb_params_{worker_priority_set_, {}};`.

I would appreciate any alternative suggestions.

Reply:
See, in most of case, main thread dispatcher is part of `ServerFactoryContext`, so, we can keep a reference in your `LoadBalancerConfig` implementation and set it when we load the config by `loadConfig`. Then your new lb implementation could get the reference from your config.

You needn't to change any exit interface or code. Right?

Reply:
@wbpcode thanks for the info.
RE the main thread dispatcher plumbing, I may be missing how `ServerFactoryContext` is passed to [loadConfig](https://github.com/envoyproxy/envoy/blob/3bf38b7b13aa884fc6c1ecade2289146ec032ef3/envoy/upstream/load_balancer.h#L281).
IIUC at the moment the `loadConfig` and the LB factory don't have access to the `ServerFactoryContext` (which is mainly used by filter-factories). Do you suggest plumbing `ServerFactoryContext` instead of the main-thread dispatcher?

Reply:
wow, I missed it, sorry.

Yeah, I think add the ServerFactoryContext as input parameter of loadConfig would be better. Just like we did in lots of other extension factory.

Reply:
It's possible and may make sense. We had that discussion internally and decided to reduce the exposure of the rest of the [ServerFactoryContext](https://github.com/envoyproxy/envoy/blob/3bf38b7b13aa884fc6c1ecade2289146ec032ef3/envoy/server/factory_context.h#L149) getters into the LB.
I don't have strong preference on either approach, and if you think it makes more sense then I think it `ServerFactoryContext` can be used.

Reply:
I personally will prefer the server factory context. But if you want to restrict the exposure of server factory context, then we can define a new context for that. Anyway, I think the main thread dispatcher should be provided by the extension factory context in most cases.

Reply:
Ack, thanks!
@efimki can you refactor the code to pass [ServerFactoryContext](https://github.com/envoyproxy/envoy/blob/3bf38b7b13aa884fc6c1ecade2289146ec032ef3/envoy/server/factory_context.h#L149) to the LB, and then use [mainThreadDispatcher()](https://github.com/envoyproxy/envoy/blob/3bf38b7b13aa884fc6c1ecade2289146ec032ef3/envoy/server/factory_context.h#L64) to fetch the dispatcher?

Reply:
Sounds good, I'll try to do that. Just to clarify, do we want to pass `ServerFactoryContext` to `TypedLoadBalancerFactory::create()` or `TypedLoadBalancerFactory::loadConfig` method?

Reply:
> PS: I think there may be a dependency problem if we use the server factory context directly. So, we may need to define a LoadBalancerFactoryContext as load balancer specific factory context.

Thanks, I've missed that before.
@efimki it seems a new LoadBalancerFactoryContext will need to be introduced. For now it can have just the mainThreadDispatcher, and we can extend it later as needed

Reply:
Sounds good to me, thank you folks!

Reply:
Done, please take a look.